Fix context.columns typescript issues

nellh opened the pull request:
Implement ColumnsMap class for preserving ordered keys when creating a parseTSV object. ColumnsMap can be constructed to return a proxy which allows the underlying map to be accessed with a property accessor (`map[key]`) like it is a regular object. A limitation of the proxy is that it only allows key access and iteration, other methods on the underlying map are not passed through and some object functions will not work.

effigies commented on bids-validator/src/types/columns.test.ts:
```diff
@@ -0,0 +1,19 @@
+    for (const [_, val] of columns) {
+      assertEquals(val, [iteration.toString()])
+      iteration += 1
+    }
```
Do we also have property access, or is it just square-bracket access `[]`?

```suggestion
    }

    // Check property access
    assertEquals(columns.a, '0')
```

Is there a way to get property access, if not?

nellh replied:
Square bracket and dot access to properties is equivalent in JavaScript but it did point out a bug with how the name is looked up. aba7edc includes a fix and your suggested test change.

effigies reviewed (approved):
I don't fully understand `columns.ts`, but everything else LGTM.

effigies commented:
Oh, we are getting a lot of:

```
 error: TypeError: context.columns.keys is not a function or its return value is not iterable
  const headers = [...context.columns.keys()]
```

Do we need to find an alternative way to do that, or can we implement `keys`?

nellh commented:
> Oh, we are getting a lot of:
> 
> ```
>  error: TypeError: context.columns.keys is not a function or its return value is not iterable
>   const headers = [...context.columns.keys()]
> ```
> 
> Do we need to find an alternative way to do that, or can we implement `keys`?

Keys can be passed from the underlying map. Let me fix that...

rwblair commented on bids-validator/src/schema/applyRules.ts:
Line 200 is failing for me:
`error: Uncaught (in promise) TypeError: context.columns.keys is not a function or its return value is not iterable`

nellh commented:
Added support for columns.keys() and a test for that but that breaks if you have a "keys" column, fixed it instead by changing to Object.keys(columns).

rwblair commented on bids-validator/src/types/columns.ts:
```diff
@@ -0,0 +1,24 @@
+// Proxy handler to implement ColumnsMapType
+export const columnMapAccessorProxy = {
+  get: function (target: ColumnsMap, prop: any) {
+    if (prop === Symbol.iterator) {
```
I had to add a `Reflect.get(target, prop)` and test return of that for typeof function then bind it to get calls to `keys` working. Not sure if there is a more elegant approach.

nellh commented:
> I can't get object.keys to work:
> 
> ```
> > import { ColumnsMap } from './columns.ts'
> undefined
> > let x = new ColumnsMap()
> undefined
> > x.a = 3
> 3
> > Object.keys(x)
> []
> ```

Fixed that with the right prototype method, we could also just add a function with a less likely to conflict name for this.
